src/engines/engine_models.rs

```diff
@@ -0,0 +1,60 @@
+//! This module provides the error enum to handle different errors associated while requesting data from
+//! the upstream search engines with the search query provided by the user.
+
+use reqwest::header::InvalidHeaderValue;
+use scraper::error::SelectorErrorKind;
+
+/// A custom error type used for handle engine associated errors.
+///
+/// This enum provides variants three different categories of errors:
+/// * `RequestError` - This variant handles all request related errors like forbidden, not found,
+/// etc.
+/// * `EmptyResultSet` - This variant handles the not results found error provide by the upstream
+/// search engines.
+/// * `UnexpectedError` - This variant handles all the errors which are unexpected or occur rarely
+/// and are errors mostly related to failure in initialization of HeaderMap, Selector errors and
+/// all other errors.
+#[derive(Debug)]
+pub enum EngineErrorKind {
+    RequestError(reqwest::Error),
+    EmptyResultSet,
+    UnexpectedError(String),
+}
+
+/// Implementing `Display` trait to make errors writable on the stdout and also providing/passing the
+/// appropriate errors that should be written to the stdout when this error is raised/encountered.
+impl std::fmt::Display for EngineErrorKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            EngineErrorKind::RequestError(request_error) => write!(f, "{}", request_error),
+            EngineErrorKind::EmptyResultSet => {
+                write!(f, "The upstream search engine returned an empty result set")
+            }
+            EngineErrorKind::UnexpectedError(unexpected_error) => write!(f, "{}", unexpected_error),
+        }
+    }
+}
+
+/// Implementing `Error` trait to make the the `EngineErrorKind` enum an error type.
+impl std::error::Error for EngineErrorKind {}
+
+/// Implementing `From` trait to map the `SelectorErrorKind` to `UnexpectedError` variant.
+impl<'a> From<SelectorErrorKind<'a>> for EngineErrorKind {
+    fn from(err: SelectorErrorKind<'a>) -> Self {
+        Self::UnexpectedError(err.to_string())
+    }
+}
+
+/// Implementing `From` trait to map the `InvalidHeaderValue` to `UnexpectedError` variant.
+impl From<InvalidHeaderValue> for EngineErrorKind {
+    fn from(err: InvalidHeaderValue) -> Self {
+        Self::UnexpectedError(err.to_string())
+    }
+}
+
+/// Implementing `From` trait to map all `reqwest::Error` to `UnexpectedError` variant.
+impl From<reqwest::Error> for EngineErrorKind {
+    fn from(err: reqwest::Error) -> Self {
+        Self::RequestError(err)
+    }
+}
```

Comment:
>For tests, I am planning to add it but would like to know your opinion like should we add all the tests at the end before we release the v1.0.0 or during every source code change??

I think we should add unit tests during every source code change.

Here is a unit test implementation.

```suggestion
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_display_request_error() {
        let error = EngineErrorKind::RequestError(reqwest::Error::new(reqwest::StatusCode::FORBIDDEN, "Forbidden"));
        assert_eq!(error.to_string(), "Request error: Forbidden");
    }
    
    #[test]
    fn test_display_empty_result_set() {
        let error = EngineErrorKind::EmptyResultSet;
        assert_eq!(error.to_string(), "The upstream search engine returned an empty result set");
    }
    
    #[test]
    fn test_display_unexpected_error_without_source() {
        let error = EngineErrorKind::UnexpectedError {
            message: String::from("An unexpected error occurred"),
            source: None,
        };
        assert_eq!(error.to_string(), "Unexpected error: An unexpected error occurred");
    }
    
    #[test]
    fn test_display_unexpected_error_with_source() {
        let source = InvalidHeaderValue::new("Invalid header value");
        let error = EngineErrorKind::UnexpectedError {
            message: String::from("An unexpected error occurred"),
            source: Some(Box::new(source)),
        };
        let expected_output = format!("Unexpected error: An unexpected error occurred\nCaused by: {}", source);
        assert_eq!(error.to_string(), expected_output);
    }
    
    #[test]
    fn test_error_source_request_error() {
        let error = EngineErrorKind::RequestError(reqwest::Error::new(reqwest::StatusCode::FORBIDDEN, "Forbidden"));
        assert_eq!(error.source().unwrap().to_string(), "Forbidden");
    }
    
    #[test]
    fn test_error_source_unexpected_error_with_source() {
        let source = InvalidHeaderValue::new("Invalid header value");
        let error = EngineErrorKind::UnexpectedError {
            message: String::from("An unexpected error occurred"),
            source: Some(Box::new(source)),
        };
        assert_eq!(error.source().unwrap().to_string(), source.to_string());
    }
    
    #[test]
    fn test_error_source_unexpected_error_without_source() {
        let error = EngineErrorKind::UnexpectedError {
            message: String::from("An unexpected error occurred"),
            source: None,
        };
        assert!(error.source().is_none());
    }
    
    #[test]
    fn test_from_selector_error() {
        let selector_error = SelectorErrorKind::Css("Invalid CSS selector");
        let error: EngineErrorKind = selector_error.into();
        assert_eq!(error.to_string(), "Unexpected error: Invalid CSS selector");
    }
    
    #[test]
    fn test_from_invalid_header_value() {
        let invalid_header_value = InvalidHeaderValue::new("Invalid header value");
        let error: EngineErrorKind = invalid_header_value.into();
        let expected_output = format!("Unexpected error: Invalid header value\nCaused by: {}", invalid_header_value);
        assert_eq!(error.to_string(), expected_output);
    }
    
    #[test]
    fn test_from_reqwest_error() {
        let reqwest_error = reqwest::Error::new(reqwest::StatusCode::FORBIDDEN, "Forbidden");
        let error: EngineErrorKind = reqwest_error.into();
        assert_eq!(error.to_string(), "Request error: Forbidden");
    }
}
```
These tests cover various scenarios, such as testing the display format of different error variants, checking the error source, and testing the From implementations for different error types.

Reply:
I think this should go in a different PR because adding too many things in one PR is never great as this makes it hard for anyone to review the project, So I would suggest putting it in a different PR altogether :slightly_smiling_face: .

Reply:
Okay, I think you're correct.